core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -1563,13 +1564,61 @@ trait EntityComponent {
       Entity(entityRecord.name, entityRecord.entityType, attributes)
 
     def unmarshalEntities(
-      entityAttributeRecords: Seq[EntityAndAttributesResult]
+      entityAttributeRecords: Seq[EntityAndAttributesResult],
+      validateOrdering: Boolean = false
     ): Seq[Entity] =
-      unmarshalEntitiesWithIds(entityAttributeRecords).map { case (_, entity) => entity }
+      unmarshalEntitiesWithIds(entityAttributeRecords, validateOrdering).map { case (_, entity) => entity }
+
+    // tail-recursive method that validates if all entity ids are contiguous within the Seq[EntityAndAttributesResult]
+    // an order of 3,3,3,2,2,2,1,1 is ok
+    // an order of 1,2,3,1 is bad, because the 1 is repeated non-contiguously
+    @tailrec
+    def validateEntityIdOrdering(prevId: Long,
+                                 current: EntityAndAttributesResult,
+                                 accum: Set[Long],
+                                 remain: Seq[EntityAndAttributesResult]
+    ): Option[EntityRecord] = {
+      if (prevId != current.entityRecord.id && accum.contains(current.entityRecord.id)) {
+        return Option(current.entityRecord)
+      }
+      if (remain.isEmpty) {
+        // we've iterated through all rows; return
+        return None;
+      }
+      // there are rows left; recurse
+      validateEntityIdOrdering(current.entityRecord.id, remain.head, accum + current.entityRecord.id, remain.tail)
+    }
 
     def unmarshalEntitiesWithIds(
-      entityAttributeRecords: Seq[EntityAndAttributesResult]
+      entityAttributeRecords: Seq[EntityAndAttributesResult],
+      validateOrdering: Boolean = false
     ): Seq[(Long, Entity)] = {
+      // TEMP: inspect how the `entityAttributeRecords` argument is ordered. We are gathering this info
+      // in prep for changing how this method works.
+
+      // short-circuit if empty.
+      if (entityAttributeRecords.isEmpty) {
+        return Seq.empty[(Long, Entity)]
+      }
+
+      if (validateOrdering) {
+        val isProperlyOrdered: Option[EntityRecord] =
+          validateEntityIdOrdering(-1, entityAttributeRecords.head, Set.empty[Long], entityAttributeRecords.tail)
+
+        if (isProperlyOrdered.isEmpty) {
```

Comment:
This led me to ask the question about return type earlier... I am confused that we're looking for `isProperlyOrdered` to be empty, but then printing a log message that says properly ordered `TRUE` -- seems like it should be the opposite case?

Reply:
Possibly rename this to `isUnordered`?

Reply:
I've updated the variable name here, good call